libc-top-half/musl/src/thread/pthread_create.c

```diff
@@ -274,14 +279,18 @@ static int start_c11(void *p)
 	return 0;
 }
 #else
-__attribute__((export_name("wasi_thread_start")))
-void wasi_thread_start(int tid, void *p)
+
+/*
+ * We want to ensure wasi_thread_start is linked whenever
+ * pthread_create is used. The following reference is to ensure that.
+ * Otherwise, the linker doesn't notice the dependency because
+ * wasi_thread_start is used indirectly via a wasm export.
+ */
+void wasi_thread_start(int tid, void *p);
+hidden void *__dummy_reference = wasi_thread_start;
+
+hidden void __wasi_thread_start_C(int tid, void *p)
 {
-	/*
-	 * Note: it's fragile to implement wasi_thread_start in C.
-	 * On entry, we don't even have C stack (__stack_pointer)
-	 * set up. Be careful when modifying this function.
-	 */
 	struct start_args *args = p;
   	__asm__(".globaltype __tls_base, i32\n"
```

Comment:
Would it make sense to move this into the asm file too and have `__wasi_thread_start_C` take tls_base as a third argument?

Reply:
IMO it's better to make the asm minimum.
unlike stack pointer, it's simple to avoid using tls in C.

Reply:
But this *is* asm code being written inline inside of C doe, which just seems more complex than just doing it in the asm file.  e.g. it depends on this rather obscure gcc asm syntax (I personally have to look this up every time).

I probably wouldn't argue for this if it wasn't the case that we are adding a new asm already and it defines a function that is the caller of this function.  Given we have a logical place to move this inline asm it makes sense to me that it live there.

Its maybe just comes down to distaste, and perhaps mistrust of the `__asm__` syntax.  IIRC, the support for inline asm in for wasm in clang is pretty limited and not very well tested.  we *kind* of made it work enough for this kind of simple case, but anything more complex and I wouldn't bet on it.

Reply:
> But this _is_ asm code being written inline inside of C code, which seems more complex than just doing it in the asm file. e.g. it depends on this rather obscure gcc asm syntax (I personally have to look this up every time).
> 
> I probably wouldn't argue for this if it wasn't the case that we are adding a new asm already and it defines a function that is the caller of this function. Given we have a logical place to move this inline asm it makes sense to me that it live there.
> 
> Its maybe just comes down to distaste, and perhaps mistrust of the `__asm__` syntax. IIRC, the support for inline asm in for wasm in clang is pretty limited and not very well tested. we _kind_ of made it work enough for this kind of simple case, but anything more complex and I wouldn't bet on it.

ok. i didn't know that the support of inline asm is limited. (do you have any references?)
i moved __tls_base stuff to asm.

Reply:
@sunfishcode might know more details as he did most of the work for make it work at all.  I don't think it ever got much very testing or attention.  IIUC until wasi-libc, emscripten was the only place it was used at all.   Basic things like this evidently work fine, which is great, but I wouldn't bet on much more.